### DISCUSSION TITLE
[Feature Proposal] Go Support for Building Native Extensions

### DISCUSSION THREAD

**Discussion by sue445** [ruby/rubygems#8128] [Ideas]
I want  `bundle gem --ext=go` in bundler. (like Rust)

# Pros
Go has some excellent features that other languages do not have, such as [goroutine](https://go.dev/tour/concurrency/1) and [Channels](https://go.dev/tour/concurrency/2).

I believe these would be more convenient if these features were available from Ruby.

Below is a benchmark comparing Ractor and goroutine.

[In my benchmark](https://github.com/sue445/go-gem-wrapper/tree/main/_benchmark), goroutine is about **x26-28 faster** than Ractor :dash:

```
$ ruby tarai.rb
go version go1.23.2 darwin/arm64
ruby 3.3.5 (2024-09-03 revision ef084cc8f4) [arm64-darwin23]
Warming up --------------------------------------
          sequential     1.000 i/100ms
   parallel (Ractor)     1.000 i/100ms
parallel (goroutine)     1.000 i/100ms
Calculating -------------------------------------
          sequential      0.019 (± 0.0%) i/s    (51.42 s/i) -      1.000 in  51.417686s
   parallel (Ractor)      0.059 (± 0.0%) i/s    (17.02 s/i) -      1.000 in  17.022087s
parallel (goroutine)      1.677 (± 0.0%) i/s  (596.20 ms/i) -      9.000 in   5.365863s

Comparison:
parallel (goroutine):        1.7 i/s
   parallel (Ractor):        0.1 i/s - 28.55x  slower
          sequential:        0.0 i/s - 86.24x  slower
```

# Details
Original idea is @mmasaki 's https://rubykaigi.org/2015/presentations/mmasaki/

At the time of his talk (2015), Go did not have a standard package manager.
So I think he had to create his own binding to call CRuby functions from Go.

However, there is now a standard package manager called [Go module](https://go.dev/doc/tutorial/create-module).

I auto-generated (almost) all of Go's bindings automatically from `ruby.h`. (about 1,100 functions)

https://github.com/sue445/go-gem-wrapper/blob/v0.1.0/ruby/function_generated.go

# What features are needed in the bundler
I made this little script to make the skeleton into a Go gem right after doing `bundle gem --ext=c`.

https://github.com/sue445/go-gem-wrapper/tree/main/_tools/patch_for_go_gem

However, I believe that if  bundler itself had equivalent functionality [like Rust](https://github.com/rubygems/rubygems/blob/v3.5.21/bundler/lib/bundler/cli/gem.rb#L207-L214), many people would be able to use this.

I would like to port what we are doing in `patch_for_go_gem.rb` to bundler if I can get the approval of others.

# Other idea
Adding functionality to the bundler core every time we add more supported languages to the `bundle gem` also seems like a high barrier.

And I think it's hard for bundler maintainers to keep maintaining many language templates...

Therefore, it might be useful to add a plugin mechanism to the `bundle gem` so that everyone can write native extension in their favorite language more casually.

e.g.  `bundle gem my_gem --template=bundler-gem-go`

bundler has a plugin mechanism.
https://bundler.io/guides/bundler_plugins.html

But it does not seem to extend `bundle gem`.

What do you all think?

**Comment by hsbt** [comment]
I'm +1 for this. Can you submit a PR for this like `go_builder.rb` under https://github.com/rubygems/rubygems/tree/master/lib/rubygems/ext?

**Reply by sue445** [reply]
@hsbt ok. I'll make a patch.

**Reply by sue445** [reply]
I created https://github.com/rubygems/rubygems/pull/8183

**Comment by sue445** [comment]
I close discussion because this features has been introduced at https://github.com/ruby/rubygems/pull/8183
